src/matrix_analysis.cu

```diff
@@ -0,0 +1,94 @@
+#include "matrix_analysis.h"
+#include <iostream>
+#include "GpuManagement.h"
+#include "kernel.h"
+#include <map>
+#include "transforms_matrix.h"
+#include "var_visitor.h"
+#include "digraph.h"
+
+void set_node_states(Matrix<float>& state, std::vector<std::shared_ptr<Node<std::set<std::string>>>>& nodes, const std::map<std::string, int>& variables){
+    for(int i = 0; i<nodes.size(); i++){
+        std::set<std::string> taint_state;
+        for (auto it = variables.begin(); it != variables.end(); it++){
+            if(state(it->second,i) != 0){
+                taint_state.insert(it->first);
+            }
+        }
+        nodes[i]->state = taint_state;
+    }
+}
+
+Matrix<float> get_initial_matrix(int var_count, int node_count){
+    Matrix<float> init_state(var_count, node_count);
+    for(int i = 0; i<node_count; i++){
+        init_state(var_count-1,i) = 1;      // set row of taint variable to 1
+    }
+    return init_state;
+}
+
+std::set<std::string> get_variables(std::vector<std::shared_ptr<Node<std::set<std::string>>>>& nodes){
+    VarVisitor<std::set<std::string>> varAnalyzer;
```

Comment:
er der ikke behov for mere meta information fordi det kan være at nogle functioner har variabler med samme navne, siden dette er et set vil variabler med samme navn have den samme taint værdi selvom de er forskellige steder i programmet

Reply:
Dette er ikke et problem, funktionen bliver primært brugt til at finde alle variabler i programmet så vi kan konstruere matricerne af den rigtige størrelse.

Derudover så behøver variabler ikke være unikke til hver funktion, da semantikken for analysen holder dem adskilt